server/src/main/java/io/druid/segment/realtime/appenderator/FiniteAppenderatorDriver.java

```diff
@@ -178,47 +185,80 @@ public void clear() throws InterruptedException
   }
 
   /**
-   * Add a row. Must not be called concurrently from multiple threads.
+   * Add a row.  This method may internally incur persisting data added so far.  Also, if too large data are persisted,
```

Comment:
@jihoonson what do you think about an alternate design:

- Change `add` to `AddResult add(InputRow row, String sequenceName, Supplier<Committer> committerSupplier)` where AddResult includes a SegmentIdentifier and something like numRowsAdded.
- Change `finish` to `publish` and have it return a `Future<SegmentsAndMetadata>` instead of `SegmentsAndMetadata`
- Make the task responsible for deciding when to call `publish` (instead of `add` like this patch has). The index task would make its decision based on numRowsAdded from AddResult.
- Rename the class from FiniteAppenderatorDriver to AppenderatorDriver. (It's no longer restricted to a finite stream!)

The idea behind this change is to make the AppenderatorDriver a good foundation for all kinds of incremental publishing, not just this one but also realtime incremental publishing like #4178. The Future return from publish is good because it prevents blocking realtime ingestion, and making the task responsible for decision about when to incrementally publish is good because the Kafka task will have some more complex logic around making sure all the replicas are in sync.